### DISCUSSION TITLE
Regarding the current function input.

### DISCUSSION THREAD

**Discussion by manjunathnilugal** [pybamm-team/PyBaMM#2010] [Q&A] [answered]
Dear all, 

I have system which generates/requires electricity along with battery charging/discharging. There are drive cycles and experimental method where I can give the charging current for '_t_' seconds to the battery in the form of the interpolant and solve.
But, is there any alternate way of giving the current function or switch the value of current from +/-  along with the running simulation with integrated pybamm model.
There is an option of _starting_solution and inital_soc_ in the solve option, I am not aware of using it, if there are any document regarding this kindly let me know.

**Comment by brosaplanella** [comment]
Not sure I fully understand your question. You can pass an interpolant (or a function) to your current. You can also work with experiments, defining "Charge" and "Discharge" steps.

You can find an example of `starting_solution` in [this notebook](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/simulating-long-experiments.ipynb).

**Comment by TomTranter** [comment, marked as answer]
I think this question relates to adjusting current input dynamically by some external control system. If this is the case then the solver can be used to step the solution for discrete time intervals with a fixed current input and this current input can be changed from step to step. Have a look at [this script](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/scripts/SPMe_step.py) to see the step solve method and [this notebook](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/change-settings.ipynb) to see how to set current as an input

**Reply by Maky55-hub** [reply]
Hi Tom,
Since we are already on the topic of using drive cycles. I wanted to ask how one can calculate the State Of Charge of the Battery. For example in the following code snippet I iterate though a list of current(postive is charge and negative current is discharge) and update the battery. How can I calculate the state of charge on each step of the iteration.
`
class PybammBattery:

    def __init__(self, capacity, soc=0):
        self.capacity = capacity
        self.soc = soc
        self.excess_power = 0
        
        self.step_solution = None
        self.step_solver = pybamm.CasadiSolver()

        # load model
        self.model = pybamm.lithium_ion.SPMe()

        # create geometry
        geometry = self.model.default_geometry

        # load parameter values and process model and geometry
        param = self.model.default_parameter_values
        param['Current function [A]'] = '[input]'
        param.process_model(self.model)
        param.process_geometry(geometry)

        # create mesh
        mesh = pybamm.Mesh(geometry, self.model.default_submesh_types, self.model.default_var_pts)

        # discretise model
        disc = pybamm.Discretisation(mesh, self.model.default_spatial_methods)
        disc.process_model(self.model)

    def update(self, current):
        if current == 0:
            return 0
        
        dt = 60
        input_parameters = {}
        input_parameters['Current function [A]'] = current
        self.step_solution = self.step_solver.step(self.step_solution, self.model, dt=dt, npts=2, inputs=input_parameters)

def simulate(battery: PybammBattery, current_list: List[float], records: List[Dict]):
    
    for current_delta in current_list:
        battery.update(current_delta)
        records.append({
            "current_delta": current_delta,
            "excess_power": battery.excess_power,
            "soc": battery.soc,
            "capacity [A.h]": battery.capacity
        })


current_list = [1, -3, 2, 3, 4, -2]
records = []  # log of some infos for later analysis

battery = PybammBattery(capacity=5)
simulate(battery, current_list, records)

result = pd.DataFrame(records)
with open("result.csv", "w") as f:
    f.write(result.to_csv())
print(result)
`

**Reply by TomTranter** [reply]
There is an example [here](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/simulating-long-experiments.ipynb). Basically you want to find the particle concentrations that correspond to the voltage limits that you cycle between and interpolate. If you use the experiment class you can't dynamically control current but maybe you don't need to and you get some summary variables without needing to do extra processing

**Reply by Maky55-hub** [reply]
Hey @TomTranter , would you please elaborate on the method you are proposing in order to calculate the SOC. If I am not wrong I need to use `step_solution` property on the model in order to calculate the SOC. I tried it but `step_solution.summary_variables` does not exist and the `step_solution.data` is also empty.

**Reply by manjunathnilugal** [reply]
Sir, I am trying to run simulation using the current and `step_solution` method.
```
import pybamm
import numpy as np
import matplotlib.pyplot as plt

model = pybamm.lithium_ion.DFN()
dt = 600
time = 0
end_time = 3600

#Loop to update c[i] after the evalutaiton of each current
while time < end_time:
    geometry = model.default_geometry
    param = pybamm.ParameterValues("Chen2020")
    param.process_model(model)
    param.process_geometry(geometry)

    # set mesh
    mesh = pybamm.Mesh(geometry, model.default_submesh_types, model.default_var_pts)

    # discretise model
    disc = pybamm.Discretisation(mesh, model.default_spatial_methods)
    disc.process_model(model)    
    
    #Current values in A which needs to updated after 600s to run simulation for 3600s
    c = np.array([3, -1.5, 2, 3, -2.2, 2.1])
    
    for i in range(len(c)):
        param.update({"Current function [A]":c[i]})
        step_solver = pybamm.CasadiSolver()
        step_solution = None
        step_solution = step_solver.step(step_solution, model, dt=dt)
        time += dt

    step_time_in_seconds = step_solution["Time [s]"].entries
    step_voltage = step_solution["Terminal voltage [V]"].entries
    plt.plot(step_time_in_seconds, step_voltage, label="SPMe (stepped solve)")
    plt.xlabel(r"$t$")
    plt.ylabel("Terminal voltage [V]")
    plt.legend()
    plt.show()
```
I am confused if I have written the code correctly or not.
Kindly let me where I am making mistake or if any mistakes I am making in the code.
Please find the attached current profile. and reference voltage behaviour
![image](https://user-images.githubusercontent.com/82285213/162032293-7ab2cfd0-ac7f-40df-8c37-784883e4dd05.png)
![image](https://user-images.githubusercontent.com/82285213/162139189-8acf044e-9569-43e2-88a0-0769fb5bae60.png)

**Reply by TomTranter** [reply]
@Maky55-hub the example I linked to contains the code you need using the ElectrodeSOH model. This gives you a couple variables which is x_100 and y_100 which are the negative and positive particle concentrations at 100% SoC which is actually determined by the voltage limits that you set in the parameter_values. You also have the zero SoC values too at x_0 and y_0. If you specify an experiment and pass that to the simulation this model gets called in the background and the Simulation also lets you specify an initial_soc so you don't have to worry about setting the initial particle concentrations. SoC is not output by the model as it's a little subjective to calculate @brosaplanella correct me if I'm wrong here. And if you add degradation mechanisms it becomes even more complicated as the capacity and internal resistances change so might the concentrations that correspond to your voltage limits (if that's what you decide to keep fixed). Here's a more worked out example of how you might calculate SoC:
```
 import pybamm
import numpy as np

parameter_values = pybamm.ParameterValues("Marquis2019")
model = pybamm.lithium_ion.SPM()
esoh_model = pybamm.lithium_ion.ElectrodeSOH()

esoh_sim = pybamm.Simulation(esoh_model, parameter_values=parameter_values)
param = model.param

# You can specify your own limits
# Original 2.8 - 4.1
Vmin = 2.8
Vmax = 4.1

parameter_values['Lower voltage cut-off [V]'] = Vmin
parameter_values['Upper voltage cut-off [V]'] = Vmax


print(f"Lower voltage cut-off [V]': {Vmin:.3f}")
print(f"Upper voltage cut-off [V]': {Vmax:.3f}")

Cn = parameter_values.evaluate(param.C_n_init)
Cp = parameter_values.evaluate(param.C_p_init)
n_Li_init = parameter_values.evaluate(param.n_Li_particles_init)

esoh_sol = esoh_sim.solve(
    [0],
    inputs={"V_min": Vmin, "V_max": Vmax, "C_n": Cn, "C_p": Cp, "n_Li": n_Li_init},
)
print(f"Initial negative electrode SOC: {esoh_sol['x_100'].data[0]:.3f}")
print(f"Initial positive electrode SOC: {esoh_sol['y_100'].data[0]:.3f}")

# Update parameter values with initial conditions
c_n_max = parameter_values.evaluate(param.c_n_max)
c_p_max = parameter_values.evaluate(param.c_p_max)

parameter_values.update(
    {
        "Initial concentration in negative electrode [mol.m-3]": esoh_sol["x_100"].data[0] * c_n_max,
        "Initial concentration in positive electrode [mol.m-3]": esoh_sol["y_100"].data[0] * c_p_max,
    }
)

c_n_min = esoh_sol["x_0"].data[0] * c_n_max
c_n_max = esoh_sol["x_100"].data[0] * c_n_max
c_p_min = esoh_sol["y_0"].data[0] * c_p_max
c_p_max = esoh_sol["y_100"].data[0] * c_p_max

print(f"Minimum negative particle concentration: {c_n_min:.3f}")
print(f"Maximum negative particle concentration: {c_n_max:.3f}")
print(f"Minimum positive particle concentration: {c_p_min:.3f}")
print(f"Maximum positive particle concentration: {c_p_max:.3f}")


sim = pybamm.Simulation(model, parameter_values=parameter_values, experiment=None)
sim.solve([0, 2*3600])

sol = sim.solution
c_n_data = sol['Average negative particle concentration [mol.m-3]'].data
c_p_data = sol['Average positive particle concentration [mol.m-3]'].data

SoC_from_n = (c_n_data - c_n_min) / (c_n_max - c_n_min)
SoC_from_p = (c_p_data - c_p_min) / (c_p_max - c_p_min)

print('SoC match', np.allclose(SoC_from_n, SoC_from_p))
```

**Reply by brosaplanella** [reply]
Yes, there are multiple definitions of SOC in the literature, so that's why at the moment we don't provide it at the standard output. However, note that most of these definitions can be easily calculated from the standard model outputs (e.g. particle surface concentration, particle average concentration...)

**Reply by TomTranter** [reply]
@manjunathnilugal here are two ways you might achieve the current control with an input:
```
import pybamm
import numpy as np
import matplotlib.pyplot as plt


plt.close('all')
pybamm.set_logging_level("INFO")

model = pybamm.lithium_ion.DFN()
dt = 600
time = 0
end_time = 3600

geometry = model.default_geometry
param = pybamm.ParameterValues("Chen2020")
Vmin = 2.8
Vmax = 4.5

param['Lower voltage cut-off [V]'] = Vmin
param['Upper voltage cut-off [V]'] = Vmax

param['Current function [A]'] = "[input]"

param.process_model(model)
param.process_geometry(geometry)

# set mesh
mesh = pybamm.Mesh(geometry, model.default_submesh_types, model.default_var_pts)

# discretise model
disc = pybamm.Discretisation(mesh, model.default_spatial_methods)
disc.process_model(model)

# Current values in A which needs to updated after 600s to run simulation for 3600s
c = np.array([3, -1.5, 2, 3, -2.2, 2.1])


step_solver = pybamm.CasadiSolver()
step_solution = None

for input_c in c:
    print(input_c)
    inputs = {'Current function [A]': input_c}
    step_solution = step_solver.step(step_solution, model, dt=dt, inputs=inputs, npts=100)

step_solution.plot()
```
and an experiment
```
import pybamm
import matplotlib.pyplot as plt

plt.close('all')
pybamm.set_logging_level("INFO")

model = pybamm.lithium_ion.DFN()
dt = 600
# param['Current function [A]'] = 3.2
param = pybamm.ParameterValues("Chen2020")
ops = [
       f"Discharge at 3.0 A for {dt} s",
       f"Charge at 1.5 A for {dt} s",
       f"Discharge at 2.0 A for {dt} s",
       f"Discharge at 3.0 A for {dt} s",
       f"Charge at 2.2 A for {dt} s",
       f"Discharge at 2.1 A for {dt} s"
       ]
experiment = pybamm.Experiment(operating_conditions=ops, period="1 s")
sim = pybamm.Simulation(model=model, experiment=experiment, parameter_values=param)
sim.solve()
sim.plot()
```

**Reply by TomTranter** [reply]
N.B you can also use `Simulation.step`

**Reply by TomTranter** [reply]
@manjunathnilugal what you had looked ok but you don't need to create all the objects inside the loop and your npts seems a bit low for the step

**Reply by manjunathnilugal** [reply]
Sir, Thanks a lot

**Reply by Maky55-hub** [reply]
Thanks a lot, appreciate the help @TomTranter @brosaplanella

**Reply by Maky55-hub** [reply]
@TomTranter @brosaplanella  Hi  again, 
[here](https://github.com/Maky55-hub/soc_pybamm_simpy/blob/main/modified_tom_model.ipynb) you will find the modified version of my code with the example code that you sent for predicting the SoC. Unfortunately on each iteration with different `current_delta` I always get the same values for `SoC_from_n` and `SoC_from_p`. Is it supposed to be like this or am I overseeing something. 
Thanks for the help
